server/src/main/java/org/elasticsearch/repositories/Repository.java

```diff
@@ -70,11 +72,24 @@ default Repository create(RepositoryMetadata metadata, Function<String, Reposito
     RepositoryMetadata getMetadata();
 
     /**
-     * Reads snapshot descriptions from the repository.
+     * Reads a collection of {@link SnapshotInfo} instances from the repository.
      *
-     * @param context get-snapshot-info-context
+     * @param snapshotIds    The IDs of the snapshots whose {@link SnapshotInfo} instances should be retrieved.
+     * @param abortOnFailure Whether to stop fetching further {@link SnapshotInfo} instances if a single fetch fails.
+     * @param isCancelled    Supplies whether the enclosing task is cancelled, which should stop fetching {@link SnapshotInfo} instances.
+     * @param consumer       A consumer for each {@link SnapshotInfo} retrieved. Called concurrently from multiple threads. If the consumer
+     *                       throws an exception and {@code abortOnFailure} is {@code true} then the fetching will stop.
+     * @param listener       If {@code abortOnFailure} is {@code true} and any operation fails then the failure is passed to this listener.
+     *                       Also completed exceptionally on cancellation. Otherwise, completed once all requested {@link SnapshotInfo}
+     *                       instances have been processed by the {@code consumer}.
      */
-    void getSnapshotInfo(GetSnapshotInfoContext context);
+    void getSnapshotInfo(
+        Collection<SnapshotId> snapshotIds,
+        boolean abortOnFailure,
+        BooleanSupplier isCancelled,
+        CheckedConsumer<SnapshotInfo, Exception> consumer,
+        ActionListener<Void> listener
+    );
```

Comment:
Was there a reason why there are both consumer and a listener rather than `ActionListener<List<SnapshotInfo>>`? Or may be with some sort of record indicating result or an individual failure if that matters?

Reply:
`SnapshotInfo` is pretty heavyweight, we need to be able to consume them incrementally rather than loading them all on-heap at once.